modules/terminal: make font configurable

t184256 opened the pull request:
Related: https://github.com/t184256/nix-on-droid-app/pull/2

@PaulGrandperrin  @Gerschtli, do you know why we haven't considered that and instead explored [copying the file over](https://github.com/t184256/nix-on-droid/issues/180)?

PaulGrandperrin commented:
I think i never actually tried doing a symlink because of this:

https://github.com/t184256/nix-on-droid/issues/120#issuecomment-867219785

> * bad news are, it has to be a regular file and not a symlink to `/nix/store` because, as far as the terminal app is concerned, there's no `/nix/store`. Auto-copying it on activation sounds hacky. Patching the app to do the sort of broken symlink resolution we already do for login script is definitely an option, but I'm basically allergic to Android development. PRs are welcome though.

PaulGrandperrin commented:
Is this working?

t184256 commented:
Seems like it is, yes =)

I guess I was wrong with my original analysis: yes, here I generate what's a broken symlink from nix-on-droid perspective, but it's a valid one from the outer point of view, and that's what matters ¯\\_(ツ)_/¯

Gerschtli commented:
Sounds legit, good idea, some minor things I found:

* Is the actual font path a gc root? Aka will it stay after a complete garbage collection?
* Should that really be `activationBefore`? Could be in normal `activation` I think.
* Should we care about keeping a backup/restore backup if there was a font installed manually before? (Maybe remove directory on unlink if empty)
* We should use `$VERBOSE_ARG` consistently in our activation scripts
* `isNull` is [deprecated](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-isNull)

Gerschtli commented on modules/terminal.nix:
```diff
@@ -0,0 +1,52 @@
+    terminal.font = mkOption {
+      default = null;
+      type = types.nullOr types.path;
+      example = "${pkgs.terminus_font_ttf}/share/fonts/truetype/TerminusTTF.ttf";
```
That will render a more usable example.
```suggestion
      example = literalExpression ''"''${pkgs.terminus_font_ttf}/share/fonts/truetype/TerminusTTF.ttf"'';
```

Gerschtli commented:
Also could be worth adding to changelog :)

t184256 commented:
Thanks for your review.

> Is the actual font path a gc root? Aka will it stay after a complete garbage collection?

It does, as the font path is embedded into the activation script.

> Should that really be activationBefore? Could be in normal activation I think.

Fixed.

> Should we care about keeping a backup/restore backup if there was a font installed manually before? (Maybe remove directory on unlink if empty)

Made it back up the font, but stopped at that. I like removing the directory even less.

> We should use $VERBOSE_ARG consistently in our activation scripts

Sprinkled.

> isNull is deprecated

Fixed.

> literalExpression

Accepted.

+ added support for both `/nix`-paths and non-mangled absolute paths.

Gerschtli reviewed (approved):
Cool, LGTM. But I am curious: Why don't you like removing empty directories created by nix-on-droid? home-manager behaves the same way.

Also, could be a good idea to log to the user, that you backuped a file.

t184256 commented:
Updated.

One more important thing I forgot to ask: does the `terminal.font` interface look fine? Without a `terminal.enabled` knob.

t184256 commented:
(will I ever get this pretty?..)

t184256 commented:
How about this? (does nothing unless there was a link; then either restores backup or removes empty dir)

Gerschtli commented:
LGTM :)

t184256 commented:
Thanks for your patience, squashing =)

MatrixManAtYrService commented:
Based on this conversation, it looks like this works.  Hurray!

I'm a bit of a noob though so like... how can I set the font?

t184256 commented:
On January 13, 2023 11:39:02 PM GMT+01:00, Matt Rixman ***@***.***> wrote:
>Based on this conversation, it looks like this works.  Hurray!
>
>I'm a bit of a noob though so like... how can I set the font?
>
>

Short story: in file `~/.config/nixpkgs/nix-on-droid.nix` set option `terminal.font`. Example: https://t184256.github.io/nix-on-droid/nix-on-droid-options.html#opt-terminal.font

MatrixManAtYrService commented:
Thank you!
